### DISCUSSION TITLE
How to install autocompletions for zsh ?

### DISCUSSION THREAD

**Discussion by 3f6a** [zellij-org/zellij#3874] [Q&A]
Following the docs I ran:
```
zellij setup --generate-completion zsh
```
which dumps a large script to my screen. Where am I supposed to place this ?

The docs only say: 
> ... append them to your shell's configuration file (...)
> See your shell's documentation for information on where to append these.

I realize someone more knowledgeable might know where to find the relevant info, but I'd appreciate some further guidance. Thanks!

**Comment by Lockszmith-GH** [comment]
## Background

Shell completion solutions have different ways of being loaded.
**zsh** specifically have elaborate competing plugin managers that deal with loading (or lazy loading) of shell extensions, as well as completion packages. (which I will not cover here)

What is common to all of these solutions, with regard to completion, is the loading of helper functions in the shell session (that's the code you saw being generated)

Each shell has a documented startup process. **zsh** starts by sourcing `$HOME/.zshrc`.

`.zshrc` will contain the loading code for every script session, this might include a plugin manager (which might load the completion code), but it can also include the completion logic directly.

The simplest way (although not necessarily the most effective one) would be to **source** (meaning loading the script's code into the session, as opposed to **running** the script).

You can (A) dump the output to a file and source that file, or (B) generate the source on the fly while the shell is initializing. (A) will be more performant, while (B) will be more dynamic (in case zellij updated, it will automatically load the 'new code').

## But what does this all MEAN ?!

If you got with (A), you will first dump the completion into a script file:

```console
$ zellij setup --generate-completion zsh > ~/zellij-completion.src
```

And then in `~/.zshrc`, at the end of the script, have the following code:
```sh
. $HOME/zellij-completion.src
```

if you go with option (B), at the end of `~/.zshrc` add the following code:
```sh
. <(zellij setup --generate-completion zsh)
```

**Reply by 3f6a** [reply]
Thanks.

I tried adding

```
. <(zellij setup --generate-completion zsh)
```

to the end of `~/.zshrc`, but I get the following error when the shell starts:

```
_arguments:comparguments:325: can only be called from completion function
/proc/self/fd/14:1229: parse error near `}'
```

**Reply by Lockszmith-GH** [reply]
ah - you're right - [that's a bug](https://github.com/zellij-org/zellij/issues/1933#issuecomment-2274464004) in one of the packages, don't think it is somehting in the zellij code, try the following code:

```zsh
autoload -U +X compinit && compinit
. <( zellij setup --generate-completion zsh | sed -Ee 's/^(_(zellij) ).*/compdef \1\2/' )
```

**Reply by Lockszmith-GH** [reply]
also, for a better method - see: [#3342](https://github.com/zellij-org/zellij/issues/3342#issuecomment-2109176801)
